vllm/model_executor/model_loader/loader.py

```diff
@@ -347,6 +347,40 @@ def load_model(self, *, model_config: ModelConfig,
                                              vision_language_config)
 
 
+class StateDictLoader(BaseModelLoader):
+
+    def __init__(self, load_config: LoadConfig):
+        super().__init__(load_config)
+        if load_config.model_loader_extra_config:
+            raise ValueError(f"Model loader extra config is not supported for "
+                             f"load format {load_config.load_format}")
+
+    def load_model(self, *, model_config: ModelConfig,
+                   device_config: DeviceConfig,
+                   lora_config: Optional[LoRAConfig],
+                   vision_language_config: Optional[VisionLanguageConfig],
+                   parallel_config: ParallelConfig,
+                   scheduler_config: SchedulerConfig) -> nn.Module:
+        from safetensors.torch import load_file
+
+        from vllm.distributed import get_tensor_model_parallel_rank
+        with set_default_torch_dtype(model_config.dtype):
+            with torch.device(device_config.device):
+                model = _initialize_model(model_config, self.load_config,
+                                          lora_config, vision_language_config)
+            rank = get_tensor_model_parallel_rank()
+            paths = glob.glob(
+                f"{model_config.model}/model-{rank}-*.safetensors")
+            params = dict(model.named_parameters())
+            for file in paths:
+                for key, val in load_file(file).items():
+                    with torch.no_grad():
+                        params[key].copy_(val)
+                    params.pop(key)
+            assert not params
```

Comment:
IIUC, unlike default loader that shards weights on the fly, the current state dict loader only supports pre-sharded checkpoints. Is it correct? If so, it would be better to support general checkpoints as StaticDictLoader can be a common loader for different use cases. Another option is adding checkers and notes:
- When `model-{rank}-*.safetensors` is missing, throw an error saying the state dict is expected to be sharded in advance.
- Add a TODO saying we should support un-sharded state dicts.

Reply:
Done - added checks and todo